src/webgpu/util/math.ts

```diff
@@ -46,9 +47,15 @@ export function diffULP(a: number, b: number): number {
   const sign_b = (u32_b & 0x80000000) !== 0;
   const masked_a = u32_a & 0x7fffffff;
   const masked_b = u32_b & 0x7fffffff;
+  const subnormal_or_zero_a = (u32_a & 0x7f800000) === 0;
+  const subnormal_or_zero_b = (u32_b & 0x7f800000) === 0;
+
+  // If the number is subnormal, then reduce it to 0 for ULP comparison.
+  const bits_a = subnormal_or_zero_a ? 0 : masked_a - 0x7fffff;
```

Comment:
I think I finally understand why the subtraction is the right thing to do here. It took me a *long* time to see it.

I think this needs another comment:

// If the number is normal then reduce its bits-representation so that we can pretend 
// that the subnormal numbers don't exist, for the purposes of counting ULP steps from
// zero (or any subnormal) to any of the normal numbers.

Reply:
Done